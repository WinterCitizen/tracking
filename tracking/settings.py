from pydantic import BaseSettings


class Settings(BaseSettings):
    jira_login: str
    jira_password: str
    jira_base_url: str

    sqlite_db_path: str
    timezone: str

    class Config:
        env_file = ".env"
