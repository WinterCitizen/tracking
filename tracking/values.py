from datetime import datetime, timedelta

from pydantic import BaseModel


class WorklogValue(BaseModel):
    issue: str
    started_at: datetime
    spent: timedelta
    comment: str

    class Config:
        frozen = True


class SavedWorklogValue(BaseModel):
    id: str
    created_at: datetime
    worklog: WorklogValue

    class Config:
        frozen = True
