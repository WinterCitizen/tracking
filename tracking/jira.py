import dataclasses
from datetime import datetime
from typing import Callable
from urllib.parse import urljoin
from zoneinfo import ZoneInfo

from aiohttp import ClientSession

from tracking.values import SavedWorklogValue, WorklogValue

# 2022-04-29T11:30:00.000+0000'

"""
'comment': 'Sync with Stan',
'started': '2022-04-29T11:30:00.000+0000',
'timeSpentSeconds': 3600,
"""


@dataclasses.dataclass
class JiraClient:
    session: ClientSession
    now: Callable[[], datetime]
    jira_base_url: str = "https://jira.clutch.co/rest/api/2/"

    async def create(self, worklog: WorklogValue) -> SavedWorklogValue:
        url = urljoin(self.jira_base_url, f"issue/{worklog.issue}/worklog")
        data = {
            "started": worklog.started_at.astimezone(ZoneInfo("UTC")).strftime("%Y-%m-%dT%H:%M:%S.000+0000"),
            "timeSpentSeconds": worklog.spent.seconds,
            "comment": worklog.comment,
        }
        async with self.session.post(url, json=data) as response:
            response.raise_for_status()
            data = await response.json()

        return SavedWorklogValue(id=data["id"], created_at=self.now(), worklog=worklog)

    async def delete(self, saved_worklog: SavedWorklogValue) -> SavedWorklogValue:
        url = urljoin(
            self.jira_base_url,
            f"issue/{saved_worklog.worklog.issue}/worklog/{saved_worklog.id}",
        )

        async with self.session.delete(url) as response:
            response.raise_for_status()

        return saved_worklog
