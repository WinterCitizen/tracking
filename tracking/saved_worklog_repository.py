import dataclasses
import logging
import sqlite3
from datetime import datetime, timedelta
from zoneinfo import ZoneInfo

import aiosqlite

from tracking.values import SavedWorklogValue, WorklogValue

logger = logging.getLogger(__name__)


@dataclasses.dataclass
class SavedWorklogRepository:
    connection: aiosqlite.Connection
    timezone: ZoneInfo

    async def write(self, saved_worklog_value: SavedWorklogValue) -> None:
        await self.connection.execute(
            """
            INSERT INTO worklogs (id, issue, started_at, spent, comment, created_at)
            VALUES (:id, :issue, :started_at, :spent, :comment, :created_at)
            """,
            {
                "id": saved_worklog_value.id,
                "issue": saved_worklog_value.worklog.issue,
                "started_at": int(saved_worklog_value.worklog.started_at.timestamp()),
                "spent": saved_worklog_value.worklog.spent.seconds,
                "comment": saved_worklog_value.worklog.comment,
                "created_at": int(saved_worklog_value.created_at.timestamp()),
            },
        )

    async def list(self, from_started_at: datetime, to_started_at: datetime) -> list[SavedWorklogValue]:
        self.connection.row_factory = sqlite3.Row
        rows = await self.connection.execute_fetchall(
            "SELECT * FROM worklogs WHERE started_at >= :from_started_at AND started_at <= :to_started_at",
            {
                "from_started_at": int(from_started_at.timestamp()),
                "to_started_at": int(to_started_at.timestamp()),
            },
        )

        saved_worklog_values: list[SavedWorklogValue] = []

        for row in rows:
            saved_worklog_values.append(
                SavedWorklogValue(
                    id=row["id"],
                    created_at=datetime.fromtimestamp(row["created_at"], tz=ZoneInfo("UTC")).astimezone(self.timezone),
                    worklog=WorklogValue(
                        issue=row["issue"],
                        started_at=datetime.fromtimestamp(row["started_at"], tz=ZoneInfo("UTC")).astimezone(
                            self.timezone,
                        ),
                        spent=timedelta(seconds=row["spent"]),
                        comment=row["comment"],
                    ),
                )
            )

        return saved_worklog_values

    async def delete(self, saved_worklog: SavedWorklogValue) -> int:
        cursor = await self.connection.execute("DELETE FROM worklogs WHERE id = :id", {"id": saved_worklog.id})
        return cursor.rowcount
