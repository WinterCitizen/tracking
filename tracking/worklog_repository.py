import dataclasses
import os
from datetime import datetime, time, timedelta
from pathlib import Path
from typing import Any
from zoneinfo import ZoneInfo

import aiofiles
import yaml
from pytimeparse import parse

from tracking.values import WorklogValue


@dataclasses.dataclass
class WorklogRepository:
    timezone: ZoneInfo

    async def read(self, worklog_path: Path) -> list[WorklogValue]:
        async with aiofiles.open(worklog_path, "r") as file:
            content = await file.read()

        track_dict = yaml.load(content, Loader=yaml.CLoader)

        date_string = os.path.basename(worklog_path).replace(".yml", "")

        start_date = datetime.strptime(date_string, "%d-%m-%Y").date()
        start_time = time.fromisoformat(track_dict["start"])

        start_datetime = datetime.combine(start_date, start_time, tzinfo=self.timezone)

        return self._parse_worklogs(start_datetime, track_dict["worklogs"])

    def _parse_worklogs(self, start_datetime: datetime, worklog_dicts: list[dict[str, Any]]) -> list[WorklogValue]:
        worklogs: list[WorklogValue] = []

        offset = timedelta()

        for worklog_dict in worklog_dicts:
            spent = timedelta(seconds=parse(worklog_dict["spent"]))

            worklogs.append(
                WorklogValue(
                    issue=worklog_dict["issue"],
                    started_at=start_datetime + offset,
                    spent=spent,
                    comment=worklog_dict["comment"],
                ),
            )

            offset += spent

        return worklogs
