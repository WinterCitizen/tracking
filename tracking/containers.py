from datetime import datetime
from functools import partial
from pathlib import Path
from typing import AsyncGenerator
from zoneinfo import ZoneInfo

import aiohttp
import aiosqlite
from dependency_injector import containers, providers

from tracking.jira import JiraClient
from tracking.saved_worklog_repository import SavedWorklogRepository
from tracking.usecases import SyncWorklogsUsecase
from tracking.worklog_repository import WorklogRepository


async def sqlite_connection_resource(
    sqlite_db_path: Path,
) -> AsyncGenerator[aiosqlite.Connection, None]:
    connection = await aiosqlite.connect(sqlite_db_path)

    yield connection

    await connection.close()


async def client_session_resource(
    auth: aiohttp.BasicAuth,
) -> AsyncGenerator[aiohttp.ClientSession, None]:
    client_session = aiohttp.ClientSession(auth=auth)

    yield client_session

    await client_session.close()


class ApplicationContainer(containers.DeclarativeContainer):
    config = providers.Configuration()

    sqlite_connection = providers.Resource(sqlite_connection_resource, sqlite_db_path=config.sqlite_db_path)
    basic_auth = providers.Factory(aiohttp.BasicAuth, login=config.jira_login, password=config.jira_password)
    session = providers.Resource(client_session_resource, auth=basic_auth)

    timezone = providers.Factory(ZoneInfo, key=config.timezone)

    saved_worklog_repository = providers.Factory(
        SavedWorklogRepository,
        connection=sqlite_connection,
        timezone=timezone,
    )

    worklog_repository = providers.Factory(WorklogRepository, timezone=timezone)

    now = providers.Factory(partial, datetime.now, tz=timezone)

    jira_client = providers.Factory(
        JiraClient,
        session=session,
        now=now,
    )

    sync_worklogs_usecase = providers.Factory(
        SyncWorklogsUsecase,
        saved_worklog_repository=saved_worklog_repository,
        worklog_repository=worklog_repository,
        jira_client=jira_client,
        timezone=timezone,
    )
