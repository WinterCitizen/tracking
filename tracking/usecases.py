import asyncio
import dataclasses
from datetime import datetime, time
from pathlib import Path
from zoneinfo import ZoneInfo

from tracking.jira import JiraClient
from tracking.saved_worklog_repository import SavedWorklogRepository
from tracking.worklog_repository import WorklogRepository


@dataclasses.dataclass
class SyncWorklogsUsecase:
    saved_worklog_repository: SavedWorklogRepository
    worklog_repository: WorklogRepository
    jira_client: JiraClient
    timezone: ZoneInfo

    async def run(self, track_path: Path) -> None:
        worklogs = await self.worklog_repository.read(track_path)

        if not worklogs:
            return

        day_start = datetime.combine(worklogs[0].started_at.date(), time.min, tzinfo=self.timezone)
        day_end = datetime.combine(day_start.date(), time.max, tzinfo=self.timezone)

        saved_worklogs = await self.saved_worklog_repository.list(day_start, day_end)

        if set(map(lambda saved_worklog: saved_worklog.worklog, saved_worklogs)) != set(worklogs):
            deleted_saved_worklogs = await asyncio.gather(
                *map(self.jira_client.delete, saved_worklogs),
                return_exceptions=True,
            )
            for deleted_saved_worklog in deleted_saved_worklogs:
                if isinstance(deleted_saved_worklog, Exception):
                    continue

                await self.saved_worklog_repository.delete(deleted_saved_worklog)

        saved_worklogs = await asyncio.gather(*map(self.jira_client.create, worklogs), return_exceptions=True)

        for saved_worklog in saved_worklogs:
            await self.saved_worklog_repository.write(saved_worklog)

        await self.saved_worklog_repository.connection.commit()
