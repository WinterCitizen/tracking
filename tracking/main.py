import asyncio
from argparse import ArgumentParser
from pathlib import Path

from dependency_injector.wiring import Provide, inject

from tracking.containers import ApplicationContainer
from tracking.settings import Settings
from tracking.usecases import SyncWorklogsUsecase


@inject
async def _main(
    sync_worklogs_usecase: SyncWorklogsUsecase = Provide[ApplicationContainer.sync_worklogs_usecase],
) -> None:
    argument_parser = ArgumentParser()
    argument_parser.add_argument("track_path", type=Path)

    namespace = argument_parser.parse_args()

    await sync_worklogs_usecase.run(namespace.track_path)


async def main():
    APPLICATION_CONTAINER = ApplicationContainer()
    APPLICATION_CONTAINER.config.from_pydantic(Settings())
    APPLICATION_CONTAINER.wire([__name__])

    await APPLICATION_CONTAINER.init_resources()

    await _main()

    await APPLICATION_CONTAINER.shutdown_resources()


if __name__ == "__main__":
    asyncio.run(main())
