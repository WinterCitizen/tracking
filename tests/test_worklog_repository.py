from pathlib import Path
from zoneinfo import ZoneInfo

from syrupy.assertion import SnapshotAssertion

from tracking.worklog_repository import WorklogRepository


async def test_worklog_repository_read(snapshot: SnapshotAssertion) -> None:
    repo = WorklogRepository(timezone=ZoneInfo("Asia/Bishkek"))

    assert await repo.read(Path("tests/test_tracks/08-05-2022.yml")) == snapshot
