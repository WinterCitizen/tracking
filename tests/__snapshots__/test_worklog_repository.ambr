# name: test_worklog_repository_read
  list([
    WorklogValue(issue='TRAC-137', started_at=datetime.datetime(2022, 5, 8, 12, 0, tzinfo=zoneinfo.ZoneInfo(key='Asia/Bishkek')), spent=datetime.timedelta(seconds=60), comment='Testerino'),
    WorklogValue(issue='TRAC-137', started_at=datetime.datetime(2022, 5, 8, 12, 1, tzinfo=zoneinfo.ZoneInfo(key='Asia/Bishkek')), spent=datetime.timedelta(seconds=180), comment='Testerino 2'),
  ])
# ---
